### DISCUSSION TITLE
Raspberry Pi Pico as .WAV-player: Ideas for performance improvements?

### DISCUSSION THREAD

**Discussion by madskjeldgaard** [pschatzmann/arduino-audio-tools#1672] [Q&A] [answered]
Hi! 

Thanks so much for making Arduino Audio Tools. I love it!

I've recently had a little project to turn a Raspberry Pi Pico board into a .wav player thing with the PCM5102A dac board. It took me some time to figure out but it's now working and fully documented in case anyone is interested:

https://github.com/madskjeldgaard/pico-audio-player-firmware

One major question though: I could not get this to work with the Pico running at normal clock frequency (130mhz) but had to overclock it to 270mhz to not have periodic dropouts in the audio being played.

Is this to be expected? Or does anyone have ideas for improving the performance of the sketch? 

Ideally I would like to play sounds at higher quality than 44.1khz/16bit if anyone's got ideas for how I can achieve that with the Pico then I would love to hear it. I tried playing 96khz/16bit files with this setup but that resulted in dropouts unfortunately.

**Comment by madskjeldgaard** [comment]
Also just generally: Any ideas on how to post the cpu usage of the Pico in this situation? On Teensy you have functions for this to help you figure out the performance but I can't find any thing for the Pico.

**Comment by pschatzmann** [comment, marked as answer]
Did you try to [identify where the bottleneck is](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working) ?
If it is the SD card you can try to increase the SPI frequency...

You can use the MeasuringStream to measure the actual thruput. 
I don't think cpu usage means anything in this context

**Reply by madskjeldgaard** [reply]
> Did you try to [identify where the bottleneck is](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working) ? If it is the SD card you can try to increase the SPI frequency...
> 
> You can use the MeasuringStream to measure the actual thruput. I don't think cpu usage means anything in this context

Great answer, thanks! 

I tried upping the SPI frequency of the SD card to 50khz and switching it out for another module type as well with no luck. 

The card I use is a Sandisk Ultra A1 Micro SD XC which works fine in a Teensy project.

Also tried with just a sine wave and that works fine without overclocking.

**Reply by pschatzmann** [reply]
50khz? The upper limit on an esp32 is 80 mhz.
You can also try to use mp3 to minimize the data amount.

ps. did you measure the max thruput by sending the output to a MeasuringStream ?

**Reply by madskjeldgaard** [reply]
Sorry that was a typo, it was 50 mhz of course. 

And no, I'll try that now. And also will try with a PWM output to see if the i2s is the issue. Thanks!

**Comment by madskjeldgaard** [comment]
After further testing I have concluded that:

- Pico 1 can play 44.1khz/16bit wave files fine, IF it is overclocked to 270mhz
- Pico 2 can go up to 96khz/16bit if it is overclocked to 300mhz.
